wm-system/src/main/kotlin/com/ajaxproject/warehouse/service/WaybillServiceImpl.kt

```diff
@@ -10,95 +10,124 @@ import com.ajaxproject.warehouse.entity.MongoProduct
+    override fun findAllWaybills(): Flux<WaybillDataLiteDto> =
+        waybillRepository.findAll().flatMap { it.mapToLiteDto() }
 
-    override fun getById(id: String): WaybillDataDto {
-        val waybill: MongoWaybill = mongoWaybillRepository.findById(ObjectId(id))
-            ?: throw NotFoundException("Waybill with id $id not found")
-        return waybill.mapToDataDto()
-    }
+    override fun getById(id: String): Mono<WaybillDataDto> =
+        waybillRepository.findById(ObjectId(id))
+            .switchIfEmpty { Mono.error(NotFoundException("Waybill with id $id not found")) }
+            .flatMap { it.mapToDataDto() }
 
     @Transactional
-    override fun updateWaybillInfo(infoUpdateDto: WaybillInfoUpdateDto, id: String): WaybillDataDto {
-        val mongoWaybill: MongoWaybill = mongoWaybillRepository.findById(ObjectId(id))
-            ?: throw NotFoundException("Waybill with id $id not found")
-        mongoCustomerRepository.findById(ObjectId(infoUpdateDto.customerId))
-            ?: throw NotFoundException("Customer with id ${infoUpdateDto.customerId} not found")
-        val updatedWaybill = mongoWaybill.setUpdatedData(infoUpdateDto)
-        return mongoWaybillRepository.save(updatedWaybill).mapToDataDto()
-    }
-
-    @Transactional
-    override fun createWaybill(createDto: WaybillCreateDto): WaybillDataDto {
-        mongoCustomerRepository.findById(ObjectId(createDto.customerId))
-            ?: throw NotFoundException("Customer with id ${createDto.customerId} not found")
-
-        val validIdList = mongoProductRepository.getValidIds(createDto.products.map { ObjectId(it.productId) })
-        createDto.products.asSequence()
-            .map { it.productId as String }
-            .minus(validIdList.toSet())
-            .map { "Product with id $it not found" }
-            .toList()
-            .takeIf { it.isEmpty() }
-            ?.let { throw NotFoundException(it) }
-
-        return mongoWaybillRepository
+    override fun createWaybill(createDto: WaybillCreateDto): Mono<WaybillDataDto> =
+        customerRepository.findById(ObjectId(createDto.customerId))
+            .switchIfEmpty { Mono.error(NotFoundException("Customer with id ${createDto.customerId} not found")) }
+            .flatMap {
+                val productIds = createDto.products.map { ObjectId(it.productId) }
+                validateProductsIds(productIds, createDto, it)
+            }.then(createWaybillAndMapToDto(createDto))
+
+    private fun validateProductsIds(
+        productIds: List<ObjectId>,
+        createDto: WaybillCreateDto,
+        it: MongoCustomer
+    ) = productRepository.getValidIds(productIds)
+        .collectList()
+        .handle<List<String>> { validIdList, sink ->
+            val notFoundIds = createDto.products
+                .asSequence()
+                .map { it.productId as String }
+                .filter { it !in validIdList }
+                .toList()
```

Comment:
What the complexity of this operation?

Reply:
O(n*m)

Reply:
Such complexity will down your DB

Reply:
it is complexity of the Kotlin algorithm, DB complexity is O(n), but anyway I have fixed it already :)